Add Support for Case Deletion

minhaminha opened the pull request:
## Product Description
<!-- For non-invisible changes, describe user-facing effects. -->

Previously, the only supported way for a user to delete a case was to either delete the user who owned that case or to archive the form that created it. This PR adds an option to directly soft delete a case and all its related subcases and submission forms, which will then be hard deleted by a periodic task 90 days later. 

This also extends the relatively new direct form deletion action. If a form a user is trying to delete was responsible for creating a case (or multiple), they'll be notified that additional cases will be deleted as well. 

Example case deletion workflow: 

1. Dedicated button for case deletion. (May move to some other location to not crowd the bottom bar).
<img width="545" alt="Screen Shot 2023-12-01 at 5 25 23 PM" src="https://github.com/dimagi/commcare-hq/assets/33491742/c70c3772-234c-4413-ab25-2e3e0f0bb39e">


2. Once you click Delete Case, you are taken to a page that details what cases (and their forms) will be deleted, what cases will be reopened due to one of those forms being archived/deleted and what cases will be affected but not deleted due to the form that updated that case being archived/deleted.  
<img width="1041" alt="Screen Shot 2023-12-01 at 5 27 09 PM" src="https://github.com/dimagi/commcare-hq/assets/33491742/3b6c2de9-6b3a-49cf-9d4f-e2df38e18ae8">

3. At the bottom of the page, you will confirm this is your intention. 
<img width="401" alt="Screen Shot 2023-12-01 at 5 30 04 PM" src="https://github.com/dimagi/commcare-hq/assets/33491742/1b6909ea-a858-4628-a0ec-842c6d3d9239">

4. After clicking the button, you'll see the final confirmation modal that asks the user to type the name of case into the input field to once again confirm that case and its related cases/forms will be deleted. 
<img width="616" alt="Screen Shot 2023-12-01 at 5 31 15 PM" src="https://github.com/dimagi/commcare-hq/assets/33491742/ddf78dc9-a531-461b-bbb0-b5c2456a4105">

To avoid attempting to delete too many cases (as some cases may have 100s of related ones from it being created by a bulk import), the amount of cases that can deleted at once will be capped at 100, though this could be changed. 
 
For more comprehensive screenshots and for an example of the form driven case deletion, there is a linked doc you can find in the ticket below (latest comment).

## Technical Summary


<!--
    Provide a link to the ticket or document which prompted this change,
    Describe the rationale and design decisions.
-->

[Jira Ticket](https://dimagi-dev.atlassian.net/browse/SAAS-14859)

Case deletion works in 3 parts:
- from the original case targeted for deletion, getting all forms associated with that case and checking if those forms have any other cases associated with it as well. This walks through all possible case relations and returns a list to reference for deletion as well as lists used to format the case confirmation page seen above. 
- Once the user submits the correct case name into the final confirmation modal, it first attempts to archive every form associated with the case/its subcases. Step 1 makes sure that the form that creates a case is the last of the forms of that case to be archived, otherwise it won't archive it. Once archived without error, it'll soft delete all forms and cases referenced.
- 90 days later those forms and cases will be picked up by the automated deletion task.

Form driven case deletion works similarly:
- If a form was responsible for creating a case (checks its case updates), the delete button on the form view page will inform the user that additional cases and their related forms will be deleted alongside the main form.
- The user will be redirected to a modified version of the case deletion page which uses the same case relations walkthrough but organizes the display data to highlight which cases were directly affected by the form being deleted and which are not.
- The user will still be prompted to type the name of a case directly created by the form before deletion. 

## Safety Assurance

### Safety story
<!--
Describe how you became confident in this change, such as
local testing, why the change is inherently safe, and/or plans to limit the blast radius of a defect.

In particular consider how existing data may be impacted by this change.
-->

Tested locally and staging. QA testing forthcoming. 

### Automated test coverage

<!-- Identify the related test coverage and the tests it would catch -->

Wrote some tests to ensure the core functions of the feature works as intended. 

### QA Plan

<!--
- Describe QA plan that along with automated test coverages proves this PR is regression free
- Link to QA Ticket
-->

TBA

### Rollback instructions

<!--
If this PR follows standards of revertability, check the box below.
Otherwise replace it with detailed instructions or reasons a rollback is impossible.
-->

- [x] This PR can be reverted after deploy with no further considerations

### Labels & Review
- [x] Risk label is set correctly
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change

minhaminha committed:
Add support for deleting individual cases (dry run only)

minhaminha committed:
Modify UI based on feedback (list structure, extra deletion confirmation modal, icon changes)

minhaminha committed:
More UI changes based on feedback (list reorganization, using list-groups, adding case icon)

minhaminha committed:
Reconstruct form name if xmlns_to_name returns the xmlns

minhaminha committed:
Add support for deleting cases and forms (no more dry run) + add tests

minhaminha committed:
Add line in task to delete old enough soft deleted cases + change wait time to 90 days before hard deleting

minhaminha committed:
Merge branch 'master' into ml/case-deletion

millerdev commented on corehq/apps/reports/standard/cases/case_data.py:
```diff
@@ -555,6 +559,209 @@ def close_case_view(request, domain, case_id):
+
+    def get_forms_for_deletion_from_case(case, subcase_count):
+        delete_cases.append(case.case_id)
+        if len(delete_cases) > MAX_CASE_COUNT or subcase_count >= MAX_SUBCASE_DEPTH:
```
Is it true that some cases will be impossible to delete because of these restrictions? Example:
- A case that was touched by a case upload, which I believe batches up to 100 cases into each form submission.
- Any case with more than 2 levels of subcases.

minhaminha replied:
As of right now, yes, the case in the first example can’t be deleted and neither can any other case that were uploaded in the same batch. I’m open to changing the value of `MAX_CASE_COUNT` though, since 10 was chosen arbitrarily. For the second example though, the case can’t be deleted but you can go into the subcases and delete those first and come back to the original case after that. I’m also open to changing the value of `MAX_SUBCASE_DEPTH`.

biyeun replied:
@millerdev do you know where we batch the 100 cases in case importer? It would be nice to make this a settings variable that can also be accessed by this case deletion code. I'm not too familiar with case importer, so couldn't find it quickly...

minhaminha replied:
so should `MAX_CASE_COUNT` be set to equal this batch size?

millerdev commented on corehq/apps/reports/standard/cases/case_data.py:
```diff
@@ -555,6 +559,209 @@ def close_case_view(request, domain, case_id):
+    def get_forms_for_deletion_from_case(case, subcase_count):
+        delete_cases.append(case.case_id)
+        if len(delete_cases) > MAX_CASE_COUNT or subcase_count >= MAX_SUBCASE_DEPTH:
+            raise ValueError("Too many cases to delete")
```
`ValueError` is a generic error that can be raised by other Python operations. It would be safer to create a custom error type to be used for this purpose. It could be as simple as
```py
class TooManyCases(ValueError):
    pass
```

millerdev commented on corehq/apps/reports/standard/cases/case_data.py:
```diff
@@ -555,6 +559,209 @@ def close_case_view(request, domain, case_id):
+                        affected_cases[touched_id][form_id] = ', '.join(actions)
+                        case_names[touched_id] = case_object.name
+            cases[case.case_id][form_id] = case_actions
+        return
```
nit: Unnecessary return statement.
```suggestion
```

millerdev commented on corehq/apps/reports/standard/cases/case_data.py:
```diff
@@ -555,6 +559,209 @@ def close_case_view(request, domain, case_id):
+        messages.error(request, _("Deleting this case would delete too many related cases. "
+                                  "Please delete some of this cases' subcases before attempting"
+                                  "to delete this case."))
+        return {}, True
```
What do you think of including the "redirect" return code in the dict?
```suggestion
        return {"redirect": True}
```

millerdev reviewed (commented):
From reading the description it sounds like it will be necessary to click three different delete buttons and enter a case name to delete a case. Could that be reduced to two delete clicks by moving the case name input to the same page that shows what forms and cases will be deleted? That seems like a better user experience to me. Has that option been considered?

This review only covers the `get_cases_and_forms_for_deletion` function. I plan to review other parts of this PR in later review sessions.

minhaminha committed:
add es_test decorator

snopoke commented on corehq/apps/reports/standard/cases/case_data.py:
```diff
@@ -555,6 +559,209 @@ def close_case_view(request, domain, case_id):
+            cases[case.case_id] = {}
+            case_names[case.case_id] = case.name
+            if len(case_names) == 1:  # only add primary label to first/main case
+                case_names[case.case_id] += ' <span class="label label-default">primary case</span>'
```
Could this be a boolean rather than raw HTML?

Similar to my comment above - perhaps a data structure would be better than a string.

snopoke commented on corehq/apps/reports/standard/cases/case_data.py:
```diff
@@ -555,6 +559,209 @@ def close_case_view(request, domain, case_id):
+    delete_forms = []  # list of forms to be soft deleted
+
+    # For formatting the list of cases/submission forms
+    cases = {}  # structured like: {case: {form: {case touched by form: actions taken by form}}}
```
perhaps a dataclass or similar data structure could be used to give this better structure.

biyeun replied:
@minhaminha I think something like a `namedtuple` would be helpful, here is an example:
https://github.com/dimagi/commcare-hq/blob/72e347c33397674503fcea10d4a67f0c1dea0a01/corehq/apps/cleanup/management/commands/undelete_docs.py#L13

you can then call `. _asdict()` on your `namedtuple`

biyeun replied:
@millerdev I always found `namedtuple` nice due to the shorter notation for a quick dataclass-esq object for simple use cases like this. curious where your preference for a full attrs object comes in vs a `namedtuple`? Usually I would use the former if I wanted optional fields and methods, but `namedtuple` for very quick named attributes

millerdev replied:
I find dataclasses and attrs to support more natural class syntax than `namedtuple`. There is tons more flexibility such as field defaults like `factory=list`, it's trivial to add methods, etc. Also, something is sticking in my mind that `namedtuple` has a performance overhead penalty, but I could be wrong about that. [SO also points out](https://stackoverflow.com/a/73724964/10840) that

>  When you compare named tuples, the names are ignored: two named tuples are equal if they contain the same values in the same order, even if they have different class names or field names

snopoke replied:
One other thing is namedtuple doesn't support default values which is often useful.

Also, I find most IDEs quite bad at backlinking to namedtuple definitions since they are just a variable and have no static class definition.

snopoke commented on corehq/apps/reports/standard/cases/case_data.py:
```diff
@@ -555,6 +559,209 @@ def close_case_view(request, domain, case_id):
+
+            case_actions = {}
+            for touched_id in touched_cases:
+                case_object = safely_get_case(request, domain, touched_id)
```
isn't this the same as `touched_cases[touched_id].case`?

minhaminha replied:
Yep - changed it to that

snopoke commented on corehq/apps/reports/standard/cases/case_data.py:
```diff
@@ -555,6 +559,209 @@ def close_case_view(request, domain, case_id):
+
+    @method_decorator(require_case_view_permission)
+    def dispatch(self, request, *args, **kwargs):
+        self.delete_dict, redirect = get_cases_and_forms_for_deletion(request, self.domain, self.case_id)
```
How long does this take to run? Wondering it would make sense to offload it to celery

minhaminha replied:
From my testing, It doesn’t take long at all. There’s also a limit on how many cases it can look at/return so the lag shouldn’t really be noticeable.

snopoke commented on corehq/form_processor/models/cases.py:
```diff
@@ -294,6 +294,21 @@ def hard_delete_cases(self, domain, case_ids, *, publish_changes=True):
 
         return deleted_count
 
+    def hard_delete_cases_before_cutoff(self, cutoff, dry_run=True):
```
Should this publish changes to Kafka to ensure the cases are removed from ES?

minhaminha replied:
The cases are removed from ES when they’re soft deleted since to the user it's supposed to look like a hard deletion (won't show up in reports).

minhaminha committed:
Make small changes based on PR feedback

minhaminha commented on corehq/apps/reports/standard/cases/case_data.py:
```diff
@@ -555,6 +559,209 @@ def close_case_view(request, domain, case_id):
+
+
+@location_safe
+def get_cases_and_forms_for_deletion(request, domain, case_id):
```
Good call - I'll refactor and tests in another commit

minhaminha commented:
> From reading the description it sounds like it will be necessary to click three different delete buttons and enter a case name to delete a case. Could that be reduced to two delete clicks by moving the case name input to the same page that shows what forms and cases will be deleted? That seems like a better user experience to me. Has that option been considered?

The final modal where you enter in the name of the case to confirm your intent to delete was meant to copy the way we currently do mobile worker deletion and I think if we were to move the input box and all the other warnings/instructions in the modal to the same page it might get a little crowded. Though I do agree it feels like a like of clicks I feel like that also emphasizes the how “dangerous” this action could be, which is something we also want. @biyeun do you have any more thoughts on reducing the amount of clicks?

biyeun commented:
@millerdev The idea behind the confirmation modal is to make it very clear to the user that this is a destructive action and they should be very sure about what they are doing. It's borrowing cues from how github handles repository deletion, or how we handle deleting a mobile worker. I think it's appropriate for this level of permanent data deletion.

minhaminha committed:
Refactor get_cases_and_forms_for_deletion in 3 separate functions + add more tests

biyeun commented on corehq/apps/reports/templates/reports/reportdata/case_delete.html:
```diff
@@ -0,0 +1,160 @@
+        The following Cases will be reopened but <strong>not</strong> deleted.
+      {% endblocktrans %}
+      <ol>
+      {% for case_link, form_link in reopened_cases.items %}
```
I see the `case_link` html is being generated in python above. can these instead just be URLs and the link HTML is specified in the template? It's always better to separate design from function.

biyeun commented on corehq/apps/reports/standard/cases/case_data.py:
```diff
@@ -555,6 +560,207 @@ def close_case_view(request, domain, case_id):
+        prepared_forms = {}
+        for form in affected_cases[case]:
+            prepared_forms[get_form_link(form)] = affected_cases[case][form]
+        prepared_affected_cases[get_case_link(case)] = prepared_forms
```
to tie in mentions of a data structure. perhaps something like
```
AffectedCase = namedtuple('AffectedCase', ['name', 'url', 'affected_forms'])
AffectedForm = namedtuple('AffectedForm', ['name, 'url', 'actions'])
```

and then in the template, you can call on the properties by iterating through a list of structured items rather than `key, value` pairs from `affected_cases.items`

minhaminha committed:
Use attrs classes instead of complex dictionaries to pass around display data

biyeun commented on corehq/apps/reports/standard/cases/case_data.py:
```diff
@@ -555,6 +561,255 @@ def close_case_view(request, domain, case_id):
+    html template. The latter will be further modified in format_case_data_for_display.
+    """
+    @define
+    class DeleteCase:
```
could these dataclasses and related utility functions live in a separate utils file outside of this report so that the dataclasses (which are looking great!) and the other related functions don't have to be nested under a larger wrapper function? It's still a bit difficult to digest `get_case_and_display_data` from top to bottom due to all the nesting. It would be great if `walk_case_relations` was further broken down too...

biyeun replied:
perhaps a better place for all these functions could be in a newly created `corehq.apps.hqcase.case_deletion_utils` file?

biyeun replied:
looks like there was also a comment earlier about adding unit tests for these extracted functions...seconding this as well! `hqcase` seems like a more appropriate place for these tests rather that `reports`, so having the utils file there will also make the organization much more digestible for future understanding

biyeun commented on corehq/apps/reports/standard/cases/case_data.py:
```diff
@@ -555,6 +561,255 @@ def close_case_view(request, domain, case_id):
+                delete_forms.insert(0, form_id)
+            form_object = XFormInstance.objects.get_form(form_id, domain)
+            if form_id not in form_names:
+                form_names[form_id] = escape(xmlns_to_name(domain, form_object.xmlns, form_object.app_id))
```
as an example guidance to further break down this function...
could all the places here where `form_names[form_id]` is being set be converted into a utility function that looks like...

```
def get_form_name(domain, form_object):
    form_name = escape(xmlns_to_name(domain, form_object.xmlns, form_object.app_id))
    if form_name == form_object.xmlns:
        inferred_name = [
            get_app(domain, form_object.app_id).name or "[Unknown App]",
            "[Unknown Module]",
             form_object.name or "[Unknown Form]"
         ]
         form_name = escape(' > '.join(inferred_name))
    return form_name
```
then right before you set `current_form`, you can do...
```
form_names[form_id] = get_form_name(domain, form_object)
```

minhaminha replied:
I ended up puling this and the process that gets the case's xforms into their own functions but no matter what I tried, I couldn't break down `walk_case_relations` further, partly because everything I tried ended up being very redundant and required passing around a lot of paramaters/context when it'd be easier to just access them in the main function. Maybe I'm thinking about this in the wrong way? 

I think even just pulling out all the dataclass definitions and those 2 functions make `get_case_and_display_data` as a whole, way more digestable. Do you have any more thoughts on this?

minhaminha committed:
Further break down get_case_and_display_data

minhaminha committed:
Disable hard deletion of eligible data

gherceg commented on corehq/apps/cleanup/tasks.py:
```diff
@@ -24,14 +24,18 @@
 
 
-@periodic_task(run_every=crontab(minute=0, hour=0), queue=getattr(settings, 'CELERY_PERIODIC_QUEUE', 'celery'))
+@periodic_task(queue=getattr(settings, 'CELERY_PERIODIC_QUEUE', 'celery'))
```
Using the `periodic_task` decorator without specifying `run_every` causes [this error](https://dimagi.sentry.io/issues/4723801186/?environment=staging&project=136860&query=is%3Aunresolved&referrer=issue-stream&statsPeriod=24h&stream_index=0) and prevents celery from properly starting up. It is probably easiest to just remove the decorator entirely, but I'd also vote for doing that on a separate PR to facilitate pushing out that change asap.

minhaminha replied:
ahh good call

minhaminha committed:
Undo delete task changes

minhaminha committed:
Integrate case deletion into form deletion if the form had created any cases

minhaminha committed:
Add more tests for form deletion, new case deletion util functions

millerdev commented:
Didn't get a chance to do a full review of these recent changes, but I have a question about this commit:

> Integrate case deletion into form deletion if the form had created any cases

Imagine a scenario with two forms, A and B, creating and updating case C (respectively). If form A is deleted (but not form B), is case C deleted?

minhaminha commented:
> Imagine a scenario with two forms, A and B, creating and updating case C (respectively). If form A is deleted (but not form B), is case C deleted?

Yep, given that A created C, if A is deleted so will C (both soft deleted). Currently, the only thing that happens to C is when A is archived (required in order to delete A) its deleted state gets set to True, which just hides from most case retrieving functions but doesn't set it up to be cleaned up later. Form B would also be deleted in this scenario for having updated C. I'll update the PR description with this new info too, since I forgot to do that yesterday.

minhaminha commented on corehq/apps/hqcase/case_deletion_utils.py:
```diff
@@ -0,0 +1,122 @@
+def get_ordered_case_xforms(case, domain):
+    # Returns deduplicated and chronologically ordered case xforms, if not already that
+    revoked_inclusive_xforms = [t.form_id for t in case.transactions if t.is_form_transaction]
+    xform_objs = [XFormInstance.objects.get_form(form_id, domain) for form_id in revoked_inclusive_xforms]
```
going to change this to to `= list({XFormInstance.objects.get_form(form_id, domain) for form_id in revoked_inclusive_xforms})` since it's going to be ordered by the received_on date anyway.

minhaminha commented on corehq/apps/reports/display.py:
```diff
@@ -143,3 +144,15 @@ def append_form_name(self, name, separator):
     return _FormType(domain, xmlns, app_id, form_name).get_label(lang, separator)
+
+
+def xmlns_to_name_case_deletion(domain, form):
```
Could probably drop the case_deletion part of the function name

millerdev commented on corehq/apps/reports/tests/test_case_data.py:
```diff
@@ -102,3 +127,283 @@ def _create_case_property(self, prop_name, group=None):
+            case_obj = CommCareCase.objects.get_case(case, self.domain)
+            self.assertTrue(case_obj.is_deleted)
+
+    def test_delete_case_that_closes_another_case(self):
```
```suggestion
    def test_delete_case_with_form_that_closes_another_case(self):
```

minhaminha committed:
Various refactors based on PR feedback

Function name changes, added description comments, added translations, added caching certain forms to reduce db hits, moved escaping to the template.

minhaminha committed:
Refactor tests + add more tests

minhaminha committed:
Refactor main case walk function into two smaller class methods, update and add tests

minhaminha commented on corehq/apps/hqcase/case_deletion_utils.py:
```diff
@@ -0,0 +1,122 @@
+            break
+    if not create_form:
+        return '[Unknown Case]'
+    case_blocks = extract_case_blocks(create_form)
```
oops forgot to respond last week but this won't be called more than once per case.

minhaminha committed:
Remove case hard deletion method (in favor of eventual tombstoning method)

minhaminha committed:
Merge branch 'master' into ml/case-deletion

millerdev commented on corehq/apps/reports/standard/cases/case_data.py:
```diff
@@ -552,7 +568,267 @@ def close_case_view(request, domain, case_id):
+        # Reorganizing display groups for form driven case deletion
+        if form_id:
+            updated_delete_cases = case_data['delete_cases']
+            for case in case_data['delete_cases']:
```
`case_data['delete_cases']` is mutated in the loop below, which could cause some items to be skipped in the iteration
```suggestion
            for case in list(case_data['delete_cases']):
```

millerdev commented on corehq/apps/reports/standard/cases/case_data.py:
```diff
@@ -552,7 +568,267 @@ def close_case_view(request, domain, case_id):
+                        self.form_delete_cases_display.append(case)
+                        updated_delete_cases.remove(case)
+                        break
+            case_data['delete_cases'] = updated_delete_cases
```
This line is unnecessary since `updated_delete_cases` references the same object as `case_data['delete_cases']`.
```suggestion
```

millerdev commented on corehq/apps/reports/standard/cases/case_data.py:
```diff
@@ -552,7 +568,267 @@ def close_case_view(request, domain, case_id):
+            case_data['delete_cases'] = updated_delete_cases
+
+            updated_affected_cases = case_data['affected_cases']
+            for case in case_data['affected_cases']:
```
```suggestion
            for case in list(case_data['affected_cases']):
```

millerdev commented on corehq/apps/reports/standard/cases/case_data.py:
```diff
@@ -552,7 +568,267 @@ def close_case_view(request, domain, case_id):
+                        self.form_affected_cases_display.append(case)
+                        updated_affected_cases.remove(case)
+                        break
+            case_data['affected_cases'] = updated_affected_cases
```
```suggestion
```

minhaminha committed:
small nit changes

minhaminha committed:
Make walk_through_case_forms always returns a dict

minhaminha committed:
Add in memory caching for forms and case blocks

millerdev commented on corehq/form_processor/models/forms.py:
```diff
@@ -836,3 +836,18 @@ def commcare_version(self):
+
+    def get_forms(self, form_ids):
+        cached_forms = [self.cache[form_id] for form_id in form_ids if form_id in self.cache]
+        get_forms = XFormInstance.objects.get_forms([form_id for form_id in form_ids if form_id not in self.cache],
+                                                    self.domain, ordered=True)
```
I don't see where these forms are added to the cache. Did I miss it?

Would it be worth adding tests?

millerdev replied:
This will hit the database with an empty list if all of the forms are already cached. It would be more efficient to skip it if there are no forms that are not cached.

minhaminha committed:
Make sure forms are actually cached in TempFormCache, remove domain argument, add test

minhaminha commented on corehq/ex-submodules/casexml/apps/case/xform.py:
```diff
@@ -226,11 +226,29 @@ def _extract_case_blocks(data, path=None, form_id=Ellipsis):
-def get_case_updates(xform, for_case=None):
+class TempCaseBlockCache:
+    def __init__(self, domain):
+        self.domain = domain
```
Removed for `TempFormCache` as well since `XFormInstance.objects.get_forms` also doesn’t use the domain that’s passed to it

minhaminha committed:
Text changes + fix escaping issue

millerdev commented on corehq/form_processor/models/forms.py:
```diff
@@ -839,15 +839,17 @@ def commcare_version(self):
+        forms = [self.cache[form_id] for form_id in form_ids if form_id in self.cache]
+        not_cached_forms = [form_id for form_id in form_ids if form_id not in self.cache]
+        if not_cached_forms:
+            retrieved_forms = XFormInstance.objects.get_forms(not_cached_forms, ordered=True)
+            for form in retrieved_forms:
+                self.cache[form.form_id] = form
+            forms += retrieved_forms
```
It doesn't look like this branch is covered by tests. What do you think of adding a test that covers it?

millerdev commented on corehq/apps/reports/templates/reports/reportdata/case_delete.html:
```diff
@@ -56,7 +57,7 @@ <h5><i class="fa fa-file"></i><a href="{{ form.url }}"> {{ form.name|escapejs }}
     {% if affected_cases %}
       {% blocktrans %}
-        The following Cases have related Form Submissions that will be deleted, which will roll back the action
+        The following case have related form submissions that will be deleted, which will roll back the action
```
This would be more grammatically correct if it said "case has" or "cases have".

millerdev commented on corehq/apps/reports/display.py:
```diff
@@ -162,5 +162,5 @@ def xmlns_to_name_for_case_deletion(domain, form):
             form.name or "[Unknown Form]"
         ]
-        form_name = ' > '.join(extracted_name)
+        form_name = escape(' > '.join(extracted_name))
```
This seems unnecessary and not the right place to be escaping for HTML. Does the template not do this escaping automatically when it renders the form name?

minhaminha replied:
Ahhh, ok so django templates do html escape by default. I agree this isn't the right place to escape but if the template escapes strings already is there a need to explicitly escape it here? I'm a little confused because I think you also suggested I do so a while back in the template itself (I used escapejs but it would escape all characters '>' characters which was a problem). Apologies for the late response also!!

millerdev replied:
I think I used poor wording in my earlier suggestion to escape in the template. The template should escape HTML characters automatically (as long as it's not tagged with `mark_safe` or some variant), so there is usually no need to use `escapejs`.

minhaminha committed:
Merge branch 'master' into ml/case-deletion

minhaminha committed:
Make small changes (remove manual escaping, add more tests about the caching objects, fix typo)

millerdev commented on corehq/ex-submodules/casexml/apps/case/tests/test_extract_caseblocks.py:
```diff
@@ -90,6 +92,37 @@ def test_blocks_in_repeat(self):
+            'case': {CASE_ATTR_ID: uuid.uuid4().hex}
+        })
+        cache_obj.cache[self.form_id] = blocks
+        block_form = create_form_for_test(self.domain, form_id=self.form_id)
```
SQL database interactions in tests are normally wrapped a transaction, which is then rolled back at the end of each test to reset the database to a clean state. However, forms and cases and a few other models are stored in "shard" databases, which do not use transactions. Therefore it is necessary to do explicit test cleanup using something like `self.addCleanup(block_form.delete)`.

millerdev commented on corehq/apps/reports/display.py:
```diff
@@ -143,3 +144,23 @@ def append_form_name(self, name, separator):
+    can't be found, this function will attempt to recreate the standard 3 part structure, rather than
+    defaulting to the xmlns url (form.xmlns). Currently only used in the case deletion workflow.
+
+    TODO: Confirm it returning the form.xmlns isn't necessary + confirm this is the format we want to display
+          for unknown form names in report tables and merge the two functions into one.
```
Will this TODO comment be addressed prior to merging this PR?

minhaminha replied:
I'll have to ask around but as of right now, I'm think about leaving it as is, since I'm not sure it's going to be as simple as I initially thought it was going to be. If I end up not doing it, I'll create a ticket for it.

millerdev commented on corehq/apps/hqcase/case_deletion_utils.py:
```diff
@@ -0,0 +1,125 @@
+def get_all_cases_from_form(form, domain, case_block_cache):
+    case_updates = get_case_updates(form, case_block_cache=case_block_cache)
+    update_ids = [update.id for update in case_updates]
+    all_cases = CommCareCase.objects.get_cases(update_ids, domain, ordered=True)
```
Would it make sense to cache these cases? In other words, is it ever possible for a given case to be fetched repeatedly by this or elsewhere in the same delete operation?

minhaminha replied:
thanks for pointing this out - I definitely thought about his while adding the form cache but forgot to!

millerdev reviewed (commented):
This is looking good. I reviewed the most recent changes and marked all of my earlier review comments as resolved in an attempt to make it easier to see if anything remains to be addressed.

minhaminha committed:
Add TempCaseCache and tests, add cleanup to tests that create forms and cases

millerdev reviewed (approved):
I think this is good to go to QA.

minhaminha committed:
Simplify test to save only once

minhaminha commented on corehq/form_processor/tests/test_cases.py:
```diff
@@ -492,6 +492,33 @@ def test_get_index_map_reversed(self):
+        same_case = CommCareCase.objects.get_case(cache_case.case_id)
+        same_case.name = "not_cached_case"
+        same_case.save()
+        retrieved_case = cache_obj.get_cases([cache_case.case_id])[0]
```
Committed this suggestion. Thank you so much for all the time you spent on this, it was extremely helpful!!

minhaminha committed:
Merge branch 'master' into ml/case-deletion

minhaminha committed:
Fix bug caused by returned form list if there is no create form

minhaminha committed:
Fix bug that raises a 404 when getting apps without an app_id

millerdev commented on corehq/apps/hqcase/case_deletion_utils.py:
```diff
@@ -0,0 +1,124 @@
+def _get_deleted_case_name(case, form_cache, case_block_cache):
+    """When a case's create form is archived, its name is reset to '', so this process sets it again
+    to properly display on the case deletion page, but does not save it to the case object"""
+    create_form = ''
+    for t in case.transactions:
+        if t.is_case_create:
+            create_form = form_cache.get_forms([t.form_id])
+            break
+    if not create_form:
+        return '[Unknown Case]'
+    case_blocks = case_block_cache.get_case_blocks(create_form[0])
```
This code is confusing to me because `create_form` is either a string or a sequence (list?), neither of which seems like a form.
```suggestion
    for t in case.transactions:
        if t.is_case_create:
            create_form = form_cache.get_forms([t.form_id])[0]
            break
    else:
        return '[Unknown Case]'
    case_blocks = case_block_cache.get_case_blocks(create_form)
```

minhaminha replied:
committed but slightly modified to avoid IndexError from sometimes empty lists returned by form_cache.get_forms, specifically when dealing with case create forms.

minhaminha committed:
Minor text and code change suggestions

millerdev commented on corehq/apps/hqcase/case_deletion_utils.py:
```diff
@@ -0,0 +1,126 @@
+def _get_deleted_case_name(case, form_cache, case_block_cache):
+    """When a case's create form is archived, its name is reset to '', so this process sets it again
+    to properly display on the case deletion page, but does not save it to the case object"""
+    for t in case.transactions:
+        if t.is_case_create:
+            form_list = form_cache.get_forms([t.form_id])
+            if form_list:
+                create_form = form_list[0]
+                break
+            continue
+    else:
+        return '[Unknown Case]'
+    case_blocks = case_block_cache.get_case_blocks(create_form)
```
Ah, I had over simplified the problem. What do you think of this solution?
```suggestion
    form_ids = [t.form_id for t in case.transactions if t.is_case_create]
    forms = form_cache.get_forms(form_ids)
    if not forms:
        return '[Unknown Case]'
    case_blocks = case_block_cache.get_case_blocks(forms[0])
```

minhaminha committed:
Fix 404 raising bug when trying to fetch deleted forms + small refactor suggestion

minhaminha committed:
Fix bug that prevents bulk imports from able to delete

millerdev commented on corehq/apps/reports/tests/test_case_data.py:
```diff
@@ -441,18 +441,36 @@ def test_case_deletion_errors_if_too_many_subcases(self):
-                CaseBlock(cases[i + 1], create=True).as_text(),
+                CaseBlock(case_id=cases[i], update={}).as_text(),
+                CaseBlock(case_id=cases[i + 1], create=True,
+                          index={'parent': (parent_type + '{}'.format(i), cases[i])}).as_text()
```
nit
```suggestion
                          index={'parent': (parent_type + str(i), cases[i])}).as_text()
```

millerdev commented on corehq/apps/reports/tests/test_case_data.py:
```diff
@@ -441,18 +441,36 @@ def test_case_deletion_errors_if_too_many_subcases(self):
+                CaseBlock(case_id=cases[i + 1], create=True,
+                          index={'parent': (parent_type + '{}'.format(i), cases[i])}).as_text()
             ], self.domain)
-        self.addCleanup(_delete_all_cases_and_forms, self.domain)
```
Was this cleanup causing a problem?

minhaminha committed:
Fix form order for large deletions

millerdev commented on corehq/apps/reports/standard/cases/case_data.py:
```diff
@@ -826,16 +826,29 @@ def post(self, request, *args, **kwargs):
+                form_instance = XFormInstance.objects.get_form(form, domain)
+                form_instance.soft_delete()
+            else:
+                raise FormArchiveError(form.form_id)
```
Raising an error hear means that some forms in `form_list` may get deleted while others (all those not processed at the time the error is raised) may not, which would leave the system in a state where (to the user) the operation appeared to fail, but some things got deleted. This does not seem like an intuitive outcome. I think the operation should either succeed with all forms deleted or fail with no forms deleted. Is that possible?

minhaminha replied:
whoops yea that wouldn't be good. I made sure all forms are archived properly first before moving onto the soft deletion (and unarchives previously archives forms if it fails in the middle of that).

millerdev replied:
Cool. Would it be hard to add a test for that functionality?

minhaminha committed:
Make form and case soft deletion all or nothing + prevent same forms from accumulating in affected cases's affected form list

minhaminha committed:
Minor fix

minhaminha committed:
Small typo fix

minhaminha committed:
Small typo fix part 2

minhaminha committed:
Reverse sorted list so latest form is archived first

minhaminha committed:
Add test covering bulk form archive error handling

minhaminha committed:
Typo fix in success message

minhaminha commented on corehq/apps/reports/tests/test_case_data.py:
```diff
@@ -487,3 +487,17 @@ def test_case_deletion_redirect_if_case_is_already_deleted(self):
+        setattr(request, 'session', {'samlSessionIndex': None})
+        setattr(request, '_messages', FallbackStorage(request))
+
+        cases, xforms = self.make_complex_case()
+        _, error = soft_delete_cases_and_forms(request, self.domain,
+                                               list(cases.values()), list(xforms.values())[:-1])
```
in making the multiple cases/forms in `make_complex_case`, the last form updates a particular child case while the second form creates it. Since on the first error it'll sort it in reverse `received_on` order, it'll at least archive the the third form before throwing an error saying it can't archive the second form, since the fourth form wasn't archived/part of the delete_form_list.

minhaminha commented:
@millerdev just a warning: there's a few more commits to come to make the delete function a background task + possibly implementing the progress bar

minhaminha committed:
Make soft_delete_cases_and_forms a celery task (needs more work)

minhaminha committed:
Merge branch 'master' into ml/case-deletion

Charl1996 commented:
@minhaminha Is this PR still on the radar?

minhaminha commented:
@Charl1996 Graham is going to be picking this back up soon

gherceg committed:
Merge branch 'master' into ml/case-deletion

minhaminha committed:
Merge branch 'master' into ml/case-deletion

gherceg committed:
Merge branch 'master' into ml/case-deletion

orangejenny commented:
Hi @minhaminha , is this PR active? I just noticed it's using requirejs for the new UIs, which is (hopefully) being deprecated soon. I can help get those switched over to webpack if that would be helpful - it shouldn't be difficult.

minhaminha commented:
@orangejenny this PR hasn't been active for almost a year now 😢 I think Graham was supposed to be taking over a while ago but it keeps getting pushed back. But feel free to make the webpack changes! No guarantees that it'll be picked back up in the near future though!!

orangejenny committed:
Switched from requirejs to webpack

minhaminha commented:
@orangejenny gotcha - thank you! :)

AmitPhulera committed:
Merge remote-tracking branch 'origin/master' into ml/case-deletion

AmitPhulera committed:
remove non-existent import
